File: workspace/public/src/native-init/s22plus_fyg8_max77705_runtime_policy_fixture.c
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include "s22plus_fyg8_max77705_result_parser.inc.c"
#include "s22plus_fyg8_max77705_envelope.inc.c"
#include "s22plus_fyg8_max77705_runtime_policy.inc.c"

static int emit_eagain(
    const char *name,
    struct s22plus_max77705_binding_witness binding,
    unsigned int expected_code) {
    uint8_t kind = 0U;
    uint8_t code = 0U;
    int rc = p316_policy_classify_eagain(&binding, &kind, &code);
    if (rc != 0 || kind != S22PLUS_MAX77705_SEMANTIC_TERMINAL
        || code != expected_code) return 1;
    printf("eagain:%s:%u:%u\n", name, kind, code);
    return 0;
}

static int emit_result(
    const char *name,
    const struct s22plus_max77705_binding_witness *binding,
    struct s22plus_max77705_runtime_result result,
    unsigned int expected_kind,
    unsigned int expected_code) {
    uint8_t kind = 0U;
    uint8_t code = 0U;
    int rc = p316_policy_classify_result(binding, &result, &kind, &code);
    if (rc != 0 || kind != expected_kind || code != expected_code) return 1;
    printf("result:%s:%u:%u\n", name, kind, code);
    return 0;
}

int main(void) {
    struct s22plus_max77705_binding_witness base = {
        .loader_state = S22PLUS_MAX77705_LOADER_RETURNED_SUCCESS,
        .pre_exact_parent_present = 1U,
        .pre_exact_parent_driver_state = S22PLUS_MAX77705_DRIVER_UNBOUND,
        .pre_matching_unbound_parent_count = 1U,
        .post_exact_parent_driver_state = S22PLUS_MAX77705_DRIVER_DIAGNOSTIC,
        .post_diagnostic_bound_parent_count = 1U,
        .post_exact_adapter_muic_0x25_client_count = 1U,
    };
    struct s22plus_max77705_binding_witness row;
    row = base;
    row.loader_state = S22PLUS_MAX77705_LOADER_IN_PROGRESS;
    row.post_exact_parent_driver_state = S22PLUS_MAX77705_DRIVER_UNBOUND;
    row.post_diagnostic_bound_parent_count = 0U;
    row.post_exact_adapter_muic_0x25_client_count = 0U;
    if (emit_eagain("probe-in-progress", row, 6U)) return 1;
    row = base;
    row.pre_exact_parent_present = 0U;
    row.pre_exact_parent_driver_state = S22PLUS_MAX77705_DRIVER_ABSENT;
    row.pre_matching_unbound_parent_count = 0U;
    row.post_exact_parent_driver_state = S22PLUS_MAX77705_DRIVER_ABSENT;
    row.post_diagnostic_bound_parent_count = 0U;
    row.post_exact_adapter_muic_0x25_client_count = 0U;
    if (emit_eagain("no-match", row, 2U)) return 1;
    row.pre_wrong_address_compatible_parent_count = 1U;
    if (emit_eagain("wrong-address", row, 3U)) return 1;
    row = base;
    row.pre_exact_parent_driver_state = S22PLUS_MAX77705_DRIVER_OTHER;
    row.post_exact_parent_driver_state = S22PLUS_MAX77705_DRIVER_OTHER;
    row.post_diagnostic_bound_parent_count = 0U;
    row.post_exact_adapter_muic_0x25_client_count = 0U;
    if (emit_eagain("other-driver", row, 4U)) return 1;
    row = base;
    row.post_exact_parent_driver_state = S22PLUS_MAX77705_DRIVER_UNBOUND;
    row.post_diagnostic_bound_parent_count = 0U;
    row.post_exact_adapter_muic_0x25_client_count = 0U;
    if (emit_eagain("post-unbound", row, 8U)) return 1;
    if (emit_eagain("bound-not-ready", base, 8U)) return 1;

    struct s22plus_max77705_runtime_result result = {0};
    result.stage = 2U;
    result.rc = -19;
    if (emit_result("identity-rejected", &base, result, 1U, 3U)) return 1;
    result.rc = -5;
    if (emit_result("identity-io", &base, result, 1U, 5U)) return 1;
    result.stage = 3U;
    result.rc = -16;
    if (emit_result("dummy-client", &base, result, 1U, 5U)) return 1;
    result.stage = 4U;
    result.rc = -5;
    if (emit_result("initial-uic", &base, result, 1U, 5U)) return 1;
    for (unsigned int stage = 5U; stage <= 9U; ++stage) {
        result.stage = (uint8_t)stage;
        result.rc = -5;
        if (stage == 8U) {
            uint8_t kind = 0U;
            uint8_t code = 0U;
            if (p316_policy_classify_result(
                &base, &result, &kind, &code) == 0) return 1;
            printf("negative:retention-stage-rejected\n");
            continue;
        }
        if (emit_result("command-transaction", &base, result, 2U, 5U))
            return 1;
    }
    result.stage = 5U;
    result.rc = 1;
    if (emit_result("positive-return", &base, result, 1U, 8U)) return 1;
    memset(&result, 0, sizeof(result));
    result.stage = S22PLUS_MAX77705_RUNTIME_STAGE_COMPLETE;
    result.response_value[0] = 0x3fU;
    result.response_value[2] = 0x09U;
    result.response_value[3] = 0x09U;
    if (emit_result("pre-nonusb-stable", &base, result, 2U, 1U)) return 1;
    result.response_value[0] = 0x09U;
    if (emit_result("pre-usb-stable", &base, result, 2U, 2U)) return 1;
    result.response_value[3] = 0x3fU;
    if (emit_result("post-reversion", &base, result, 2U, 3U)) return 1;
    result.response_value[2] = 0x3fU;
    if (emit_result("complete-other", &base, result, 2U, 4U)) return 1;

    static const char canonical[] =
        "v=1 stage=10 rc=0 pmic_v=03 pmic_id=15 pmic_rev=02 "
        "uic0_v=1 uic0=04 issued=0d seen=0d wr_attempt=0 wr_amb=0 "
        "rsp=05000505 val=09000909 p0n=1 p0=80 p1n=0 p1= "
        "p2n=1 p2=80 p3n=1 p3=80\n";
    struct s22plus_max77705_runtime_poll_summary summary;
    if (s22plus_max77705_runtime_parse_result(
        canonical, sizeof(canonical) - 1U, &result, &summary) != 0) return 1;
    uint8_t envelope[128];
    uint16_t detail = 0U;
    if (s22plus_max77705_encode_envelope(
        &base, S22PLUS_MAX77705_SEMANTIC_MUX,
        S22PLUS_MAX77705_MUX_PRE_USB_STABLE_USB,
        S22PLUS_MAX77705_OBSERVER_SITE_NONE,
        S22PLUS_MAX77705_OBSERVER_ERROR_NONE,
        &result, &summary, envelope, &detail) != 0 || detail != 0x6711U)
        return 1;
    return 0;
}

File: workspace/public/src/scripts/revalidation/s22plus_fyg8_max77705_runtime_policy_fixture.py
#!/usr/bin/env python3
"""Execute the exact native P3.16 EAGAIN/result classification policy."""

from __future__ import annotations

import hashlib
import json
from collections import Counter
from pathlib import Path
import subprocess
import tempfile


SCHEMA = "s22plus_fyg8_max77705_runtime_policy_fixture_v1"
VERDICT = "PASS_S22PLUS_FYG8_MAX77705_RUNTIME_POLICY_HOST_ONLY"
SOURCE = Path(
    "workspace/public/src/native-init/"
    "s22plus_fyg8_max77705_runtime_policy_fixture.c"
)
INCLUDE = Path("workspace/public/src/native-init")
CLANG = Path(
    "workspace/private/work/toolchains/aosp-clang-android12-release/"
    "clang-r416183b/bin/clang"
)


class FixtureError(ValueError):
    pass


def audit(root: Path) -> dict[str, object]:
    source = root / SOURCE
    compiler = root / CLANG
    if not source.is_file() or not compiler.is_file():
        raise FixtureError("runtime policy fixture input is missing")
    with tempfile.TemporaryDirectory(prefix="s22-max77705-policy-") as name:
        directory = Path(name)
        executable = directory / "fixture"
        command = [
            str(compiler), "-std=c11", "-O2", "-Wall", "-Wextra", "-Werror",
            "-I", str(root / INCLUDE), str(source), "-o", str(executable),
        ]
        built = subprocess.run(
            command, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
            timeout=60, check=False,
        )
        if built.returncode != 0:
            raise FixtureError(f"runtime policy compile failed: {built.stderr!r}")
        ran = subprocess.run(
            [str(executable)], stdout=subprocess.PIPE, stderr=subprocess.PIPE,
            timeout=10, check=False,
        )
        if ran.returncode != 0 or ran.stderr:
            raise FixtureError(
                f"runtime policy execution failed: rc={ran.returncode} "
                f"stderr={ran.stderr!r}"
            )
        lines = ran.stdout.decode("ascii").splitlines()
        expected = Counter(
            {
                "eagain:probe-in-progress:1:6": 1,
                "eagain:no-match:1:2": 1,
                "eagain:wrong-address:1:3": 1,
                "eagain:other-driver:1:4": 1,
                "eagain:post-unbound:1:8": 1,
                "eagain:bound-not-ready:1:8": 1,
                "result:identity-rejected:1:3": 1,
                "result:identity-io:1:5": 1,
                "result:dummy-client:1:5": 1,
                "result:initial-uic:1:5": 1,
                "result:command-transaction:2:5": 4,
                "negative:retention-stage-rejected": 1,
                "result:positive-return:1:8": 1,
                "result:pre-nonusb-stable:2:1": 1,
                "result:pre-usb-stable:2:2": 1,
                "result:post-reversion:2:3": 1,
                "result:complete-other:2:4": 1,
            }
        )
        if Counter(lines) != expected:
            # Five command stages deliberately share one retained semantic.
            raise FixtureError("runtime policy row geometry differs")
        cross = directory / "cross.c"
        cross.write_text(
            "#include <stddef.h>\n#include <stdint.h>\n"
            "void *memset(void *p,int v,size_t n){unsigned char*q=p;"
            "while(n--)*q++=(unsigned char)v;return p;}\n"
            f'#include "{root / INCLUDE / "s22plus_fyg8_max77705_result_parser.inc.c"}"\n'
            f'#include "{root / INCLUDE / "s22plus_fyg8_max77705_envelope.inc.c"}"\n'
            f'#include "{root / INCLUDE / "s22plus_fyg8_max77705_runtime_policy.inc.c"}"\n'
            "int policy_result(struct s22plus_max77705_binding_witness*b,"
            "struct s22plus_max77705_runtime_result*r,unsigned char*k,"
            "unsigned char*c){return "
            "p316_policy_classify_result(b,r,k,c);}\n"
            "int policy_eagain(struct s22plus_max77705_binding_witness*b,"
            "unsigned char*k,unsigned char*c){return "
            "p316_policy_classify_eagain(b,k,c);}\n"
            "int policy_parse(const char*p,size_t n,"
            "struct s22plus_max77705_runtime_result*r,"
            "struct s22plus_max77705_runtime_poll_summary*s){return "
            "s22plus_max77705_runtime_parse_result(p,n,r,s);}\n"
            "int policy_encode(struct s22plus_max77705_binding_witness*b,"
            "struct s22plus_max77705_runtime_result*r,"
            "struct s22plus_max77705_runtime_poll_summary*s,"
            "unsigned char e[128],unsigned short*d){return "
            "s22plus_max77705_encode_envelope(b,2,1,0,0,r,s,e,d);}\n",
            encoding="ascii",
        )
        crossed = subprocess.run(
            [
                str(compiler), "--target=aarch64-linux-gnu", "-std=c11",
                "-ffreestanding", "-fno-builtin", "-Wall", "-Wextra",
                "-Werror", "-c", str(cross), "-o", str(directory / "cross.o"),
            ],
            stdout=subprocess.PIPE, stderr=subprocess.PIPE,
            timeout=60, check=False,
        )
        if crossed.returncode != 0:
            raise FixtureError(
                f"runtime policy AArch64 compile failed: {crossed.stderr!r}"
            )
    return {
        "schema": SCHEMA,
        "verdict": VERDICT,
        "eagain_rows": 6,
        "result_rows": 14,
        "unique_rows": 17,
        "shared_transaction_rows": 4,
        "negative_invariants": 1,
        "native_stdout_sha256": hashlib.sha256(ran.stdout).hexdigest(),
        "aarch64_freestanding_compile": True,
        "verified": True,
    }


def main() -> int:
    root = Path(__file__).resolve().parents[5]
    print(json.dumps(audit(root), sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
